File: chazm-model/src/main/java/runtimemodels/chazm/model/entity/InstanceGoalFactory.java
package runtimemodels.chazm.model.entity;

import runtimemodels.chazm.api.entity.InstanceGoal;
import runtimemodels.chazm.api.entity.SpecificationGoal;
import runtimemodels.chazm.api.id.UniqueId;

/**
 * The {@linkplain InstanceGoalFactory} interface defines the APIs for constructing {@linkplain InstanceGoal}s.
 *
 * @author Christopher Zhong
 * @since 7.0.0
 */
@FunctionalInterface
public interface InstanceGoalFactory {

    /**
     * Constructs an {@linkplain InstanceGoal}.
     *
     * @param id        the {@linkplain UniqueId} that represents the {@linkplain InstanceGoal}.
     * @param goal      the {@linkplain SpecificationGoal} of the {@linkplain InstanceGoal}.
     * @param parameter the {@linkplain InstanceGoal.Parameter} of the {@linkplain InstanceGoal}.
     * @return an {@linkplain InstanceGoal}.
     */
    InstanceGoal buildInstanceGoal(UniqueId<InstanceGoal> id, SpecificationGoal goal, InstanceGoal.Parameter parameter);

}

File: chazm-model/src/main/java/runtimemodels/chazm/model/entity/InstanceGoalFactory.kt
package runtimemodels.chazm.model.entity

import runtimemodels.chazm.api.entity.InstanceGoal
import runtimemodels.chazm.api.entity.SpecificationGoal
import runtimemodels.chazm.api.id.UniqueId

/**
 * The [InstanceGoalFactory] interface defines the APIs for constructing [InstanceGoal]s.
 *
 * @author Christopher Zhong
 * @since 7.0.0
 */
@FunctionalInterface
interface InstanceGoalFactory {

    /**
     * Constructs an [InstanceGoal].
     *
     * @param id        the [UniqueId] that represents the [InstanceGoal].
     * @param goal      the [SpecificationGoal] of the [InstanceGoal].
     * @param parameter the [InstanceGoal.Parameter] of the [InstanceGoal].
     * @return an [InstanceGoal].
     */
    fun buildInstanceGoal(id: UniqueId<InstanceGoal>, goal: SpecificationGoal, parameter: InstanceGoal.Parameter): InstanceGoal

}
